frontend/src/app/components/channel/small/small.controller.ts

```diff
@@ -0,0 +1,204 @@
+      return false;
+    }
+
+    return true;
+  }
+
+  private acceptUser(login: string) {
+    // Re-forward preview user set to SDI in case of change
+    this.constructUserSets();
+
+    if (this.userSetIndex.preview === this.userSets.length - 1) {
+      this.putUserSetToPreview(this.userSetIndex.preview);
+    }
+  }
+
+  private reforwardSlotOnUserRemoval(changedUserSetIndex: number, program: boolean): ng.IPromise<any> {
+    var slotName = this.getSlotName(program);
+    var deffered = this.$q.defer();
+
+    if (this.userSetIndex[slotName] > this.userSets.length) {
+      // User set removed
+      return this.putUserSetToSlot(null, program);
+    } else if (changedUserSetIndex === this.userSetIndex[slotName]) {
+      // Slot user set changed
+      return this.putUserSetToSlot(changedUserSetIndex, program);
+    }
+
+    deffered.resolve();
+    return deffered.promise;
+  }
+
+  private putUserSetToSlot(index: number, program: boolean): ng.IPromise<any> {
+    var deffered = this.$q.defer();
+
+    var slotName = this.getSlotName(program);
+
+    if (index === this.userSetIndex[slotName]) {
+      deffered.resolve();
+    }
+
+    var userSet: IUser[] = [];
+
+    if (index !== null) {
+      userSet = this.userSets[index];
+    }
+
+    this.userSetIndex[slotName] = index;
+    this.isForwarded[slotName] = false;
+
+    var videoPorts = this.config.janus.sdiPorts[this.name].video[slotName];
+
+    var logins: string[] = Array.apply(null, Array(this.userSetSize));
+
+    userSet.forEach((user: IUser, userIndex: number) => {
+      logins[userIndex] = user.login;
+    });
+
+    // TODO: move to base controller
+    this.videoRoom.forwardRemoteFeeds(logins, videoPorts).then(() => {
+      this.isForwarded[slotName] = true;
+
+      // Forwarding succeeded, changing titles
+      logins.forEach((login: string, loginIndex: number) => {
+        var title: string;
+
+        if (this.usersByLogin[login]) {
+          title = this.usersByLogin[login].title;
+        } else {
+          title = '';
+        }
+
+        this.videoRoom.changeRemoteFeedTitle(title, videoPorts[loginIndex]);
+
+        deffered.resolve();
+      });
+    }, () => {
+      var error = 'Failed forwarding feed to SDI';
+      this.toastr.error(error);
+      console.error(error);
+      deffered.reject();
+    });
+
+    return deffered.promise;
+  }
+
+  private getSlotName(program: boolean): string {
+    var slotName = program ? 'program' : 'preview';
+    return slotName;
+  }
+
+  private getUserSetIndexForUser(login: string): number {
+    var user = this.usersByLogin[login];
+
+    this.userSets.forEach((userSet: IUser[], userSetIndex: number) => {
+      if (userSet.indexOf(user) !== -1) {
+        return userSetIndex;
+      }
+    });
+
+    return null;
+  }
+
+  private constructUserSets(): void {
+    this.userSets = [];
+
+    var userSet: IUser[] = [];
+    var onlineUsers = this.getOnlineUsers();
```

Comment:
We want the algorithm for changing the fours to make more sense for the operator. Please consult Micha. 

On  each change in user state (joined or left) we want only one four to change and not all of them.